### PR TITLE
Support bash-completion

### PR DESCRIPTION
**Pull request by nak3**
This PR adds bash-completion support.
Running `$./hack/run-gendoc.sh` will create bash-completion scripts  `rel-eng/completions/bash/{openshift,osc,osadm}`

### CODE REVIEW

**Comment by smarterclayton** [comment]
Can you split this into a few commits?  The first should be the core code that can generate docs, the second should be the changes necessary to cleanup the docs/scripts/completions, and the third should be any of the generated code.  Thanks

**Comment by nak3** [comment]
Sure, I did it.

**Comment by smarterclayton** [comment]
Handing off to Fabiano for review

**Comment by fabianofranz** [comment]
Reviewing, @eparis you may want to jump in too.

File: rel-eng/completions/bash/openshift [hunk 1]
```diff
@@ -0,0 +1,3146 @@
...
+
+    flags=()
+    two_word_flags=()
+    flags_with_completion=()
+    flags_completion=()
+
+    flags+=("--help")
+    flags+=("-h")
+
+    must_ha
```
Comment (eparis) [inline rel-eng/completions/bash/openshift]: Does this work?  Isn't the command name "osc"   ?  Where does this full "openshift" name come from?

File: rel-eng/completions/bash/openshift [hunk 2]
```diff
@@ -0,0 +1,3146 @@
...
+    commands=()
+
+    flags=()
+    two_word_flags=()
+    flags_with_completion=()
+    flags_completion=()
+
+    flags+=("--filename=")
+    flags_with_completion+=("--filename")
+    flags_completion+=("_filedir '@(json|yaml|yml)'")
```
Comment (eparis) [inline rel-eng/completions/bash/openshift]: This doesn't work.

Comment (eparis) [inline rel-eng/completions/bash/openshift]: See:
https://github.com/spf13/cobra/commit/bba56042cf767e329430e7c7f68c3f9f640b4b8b
https://github.com/GoogleCloudPlatform/kubernetes/commit/930f8645b591a9a09f46cea6ed7a722786bc1f58

File: pkg/cmd/cli/describe/describer.go [hunk 3]
```diff
@@ -28,55 +28,52 @@ import (
...
+		"ClusterRoleBinding":   &ClusterRoleBindingDescriber{c},
+		"ClusterRole":          &ClusterRoleDescriber{c},
+		"User":                 &UserDescriber{c},
+		"UserIdentityMapping":  &UserIdentityMappingDescriber{c},
+	}
+	return m
+}
+
+// List of all resource types we can describe
+func DescribableResources(c *client.Client, kclient kclient.Interface, host string) []string {
```
Comment (deads2k) [inline pkg/cmd/cli/describe/describer.go]: args are dead?

**Comment by nak3** [comment]
I included this request [pull/8039](https://github.com/GoogleCloudPlatform/kubernetes/pull/8039)

Now I think I fixed all reviewed points.
Please let me know if I can squash this PR.

File: rel-eng/completions/bash/osc [hunk 4]
```diff
@@ -0,0 +1,1091 @@
...
+        return 1
+    fi
+    __osc_parse_get ${nouns[${#nouns[@]} -1]}
+}
+
+# $1 is the name of the pod we want to get the list of containers inside
+__osc_get_containers()
+{
+    local template
+    template="{{ range .desiredState.manifest.containers  }}{{ .name }} {{ end }}"
```
Comment (eparis) [inline rel-eng/completions/bash/osc]: Isn't this still broken?  Move to v1beta3 template form and LGTM

**Comment by nak3** [comment]
@eparis Thank you so much.
@smarterclayton @fabianofranz I squash all commits, but if you prefer split commits, please let me know. I will split this again.

**Comment by smarterclayton** [comment]
LGTM [merge]

**Pull request merged**